### PR TITLE
Split platform targets

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
platform_device has been split into platform_runtime and platform_LA.
This is consistent with most software stacks that runtime library and linear algebra libraries are separated.
`src/Platforms/CMakeLists.txt` has details about how to use these two CMake targets.
This MR also reduces the amount of object targets.

## What type(s) of changes does this code introduce?
- Build related changes

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server, bora, spock, summit

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Comment by ye-luo** [comment]
test this please

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
Overall I think this is a good change and want to get it in fast.  A few issues though.

First what is going on with the containers target in this PR disappearing from target_link_libraries calls seems outside the scope of the PR.
But that seems out of scope for this PR.  Maybe just remove the changes that only change by removing the containers target.
I started an issue for this #3441 

Also I want to clarify the direction blas/lapack support is going in. Right now we name spacing around the different blas and lapack flavors and this is very useful since there is code where you need to know for sure and control which of these you call.  I don't think this should without real design work become a way to have generic blas/lapack calls go to device specific calls that is controlled by a compile time switch. The concerns of different blas platforms are such that I think such a thing is very premature.  Don't care if it would "work" for OpenMP and OMP target pragmas getting turned off and on.

**Comment by ye-luo** [comment]
`become a way to have generic blas/lapack calls go to device specific calls` this is never the direction we are moving forward. Calls should carry namespace (BLAS(CPU), ompBLAS, cuBLAS..)

CMake target is different from source code, when platform_LA is linked, it supplies all the BLAS libraries (include path, library files). That is it.

**Comment by ye-luo** [comment]
> First what is going on with the containers target in this PR disappearing from target_link_libraries calls seems outside the scope of the PR.

Since CMake is not very restrictive. If you don't know qmcutils depends on containers publicly. having `A depends on qmcutil and contains` is forgiving.
If you know qmcutils depends on containers and it can be handled by CMake, leaving the work to CMake is preferred. In this way, we keep minimal forced dependency.

File: src/Platforms/CMakeLists.txt [hunk 1]
```diff
@@ -2,55 +2,63 @@
...
-# We will move to functionality based targets
-# platform_runtime handles device management, memory management
-# add_library(platform_runtime INTERFACE)
-# platform_lapack handles BLAS/LAPACK
-# add_library(platform_lapack INTERFACE)
-
-set(HOST_SRCS Host/sysutil.cpp Host/InfoStream.cpp Host/OutputManager.cpp)
+# Targets under Platforms are designed to abstract external libraries
+# platform_runtime, platform_LA may be used outside Platforms.
+# Device specific targets should not be referenced used outside Platforms.
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/Platforms/CMakeLists.txt:15]: I would say it depends.
For a test under QMCWaveFunctions, if your test link qmcwfs, then you should never need to link platforms directly.
If you are working on a test for DiracDeterminantBatched, then still better to link platform_LA. Only when you only need to test CUDA, then you can link platform_cuda_LA but we may need to add INCLUDE_DIRECTORIES I have not tested it well. We can improve this for such tests. In general, I think this is convenience vs minimal dependency. Different trade off. If you have a case, we can discuss how to improve it after this PR landed.

**Review by PDoakORNL — APPROVED** [review]

File: src/Particle/ParticleBase/tests/CMakeLists.txt [hunk 2]
```diff
@@ -19,7 +19,7 @@ add_executable(${UTEST_EXE} test_particle_attrib.cpp test_random_seq.cpp test_at
 use_fake_rng(${UTEST_EXE})
 target_link_libraries(${UTEST_EXE} catch_main qmcparticle qmcfakerng)
 if(USE_OBJECT_TARGET)
-  target_link_libraries(${UTEST_EXE} qmcutil containers platform_omptarget)
+  target_link_libraries(${UTEST_EXE} qmcutil)
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/Particle/ParticleBase/tests/CMakeLists.txt:22]: Agreed. qmcutil is bad catchall target.
In this particular line, it is only mean to make USE_OBJECT_TARGET for Clang offload.

> If a test depends just on containers and platform don't link qmcutil.

Is this not the case today?

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/Particle/ParticleBase/tests/CMakeLists.txt:22]: Yes I didn't follow the reason for this removal.

**Comment by ye-luo** [comment]
Start testing in-house

**Pull request merged**